### ISSUE TITLE
(partial) update of documents

### ISSUE THREAD

**Issue by colinsanders** [logstash-plugins/logstash-output-elasticsearch#114]
Hi, 
I would like to be able merge  documents from different sources with the same "document_id" with the update API or create a new document ("doc_as_upsert": true) is it does not exist yet.
In 1.4.2. I changed the elasticsearch_http.rb slightly to do this. 
Several sources says that it should would work with the  protocol => "http".
But I can't make it work in logstash-1.5.0.rc2. 
Any suggestions? help? 
Thanks,
Colin

**Comment by Da-Wei** [comment]
Hi,
I need it also, i work on a branch (https://github.com/Da-Wei/logstash-output-elasticsearch/tree/update_action) with support for node/http protocol and scripting.

scripting doesn't work correctly, but for doc updates, it's fine : 
output {
  elasticsearch {
    action => "update"
    document_id => "%{[mydocid]}"
    doc_as_upsert => true
  }
}
or with specific upsert : 
output {
  elasticsearch {
    action => "update"
    document_id => "%{[mydocid]}"
    upsert => {
       value => "%{[myvalue]}"
    }
  }
}

Hope this can help !
Comments are welcomed
David.

**Comment by colinsanders** [comment]
Hi David,

Many thanks for your reply, 
I did some (small) tests with the http protocol and it works well. 
I hope the logstash team will integrate it into the master.

I continue testing and if I notice something I will let you now.

Merci beaucoup.

Colin

**Comment by talevy** [comment]
related: #16 

@Da-Wei awesome! do you mind opening a PR with these changes. I was going to go ahead and add this support in response to #16. Looks like you're one step ahead! it would be awesome to get this into master soon, I'm glad to help out with testing as well.

**Comment by Da-Wei** [comment]
Yes @tavely. I just have to split my branch in two others : one for document update/upsert and one for scripting update.
scripting update is more complicated (for node protocol) : need java dependencies for scripting languages like groovy and others problems like how to pass a variable like '@timestamp' to a language which doesn't support '@' for first char

I hope to provide the first PR today ...

**Comment by talevy** [comment]
awesome, looking forward to it!

**Comment by cdahlqvist** [comment]
For scripted updates, would it perhaps make sense to at least initially limit the support to executing named scripts on the server as this is more secure and also should reduce the amount of dependencies?

**Comment by Da-Wei** [comment]
@cdahlqvist, it seems that when using node protocol, script interpretation is done locally, not on remote nodes, so in this mode we need to add scripting engine (groovy.jar, ...) to the classpath of logstash (or embedded it with the es plugin) and our script(s) (for named scripts) in [logstash workdir]/config/scripts/
For others protocols, no problems : it's the same api for script or named script and security is done on elasticsearch' side, isn't it ?

**Comment by headiron** [comment]
Hi David, Colin.

I also need plugin for partial update.
By spending lots of time, I finally success build this plugin.
But I can't find way how I could install plugin to my logstash ? 

Could you let me know how I could install plugin at logstash ? 

Best Regards
Ducheol

**Comment by Da-Wei** [comment]
Hi @headiron,
Check pull request #116,
Checkout my branch 
```git clone -b update_and_upsert https://github.com/Da-Wei/logstash-output-elasticsearch.git```
in logstash-output-elasticsearch directory, execute :
```gem build logstash-output-elasticsearch.gemspec # this command produce logstash-output-elasticsearch-0.2.3-java.gem)```
Then from your logstash installation (need >=1.5.0) install the produced gem :
```plugin install logstash-output-elasticsearch-0.2.3-java.gem```

David.

**Comment by headiron** [comment]
Hi David.

Thanks for the help. Now, I got update plugin and work well as you explained.
In my problem, I clone the trunk and try to install logstash 1.4.2

Best Regards
Ducheol

**Comment by JuPyoHong** [comment]
awesome! thx Da-Wei! :p

**Comment by JuPyoHong** [comment]
@Da-Wei 
hi david. first, i can speak/write english a little.. sorry

I use your output plugin. it is nice.. thx. but  when i try upgrade your plugin to 0.2.9, occur error that not found 'update' action. what happen..? please help me.

and, 
'Got error to send bulk of actions to elasticsearch server at localhost : Read timed out '
this error is original plugin issue?? this error is occur in your plugin 0.2.3

**Comment by Da-Wei** [comment]
@JuPyoHong, I rebase the branch update_and_upsert with master two weeks ago and I have tested the same method as my previous comment and all is fine. Could you explain your install context ?

For the "read timed out" see #103. I think that it's resolved with v0.2.9.

**Comment by sbeaupre** [comment]
@Da-Wei are you still working on the scripting support? (even if it is only for http initially)

**Comment by Da-Wei** [comment]
Yes, @sbeaupre.  I just restart to work on it.

**Comment by sbeaupre** [comment]
@Da-Wei Good to hear! I tried to adapt the 1.0.7 version with your modification you made earlier: https://github.com/logstash-plugins/logstash-output-elasticsearch/commit/17b583ed7c8f15f072dcc6aea050c2d61800a1d0

but I keep on getting following error (with --debug):

```
Plugin not defined in namespace, checking for plugin file {:type=>"output", :name=>"elasticsearch", :path=>"logstash/outputs/elasticsearch", :level=>:debug, :file=>"logstash/plugin.rb", :line=>"133", :method=>"lookup"}
[2015-08-13 22:38:00.344]  WARN -- Concurrent: [DEPRECATED] Java 7 is deprecated, please use Java 8.
Java 7 support is only best effort, it may not work. It will be removed in next release (1.0).
The error reported is: 
  undefined method `version' for nil:NilClass
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.2.2-java/lib/logstash/util/plugin_version.rb:26:in `find_version!'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.2.2-java/lib/logstash/util/plugin_version.rb:36:in `find_plugin_version!'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.2.2-java/lib/logstash/config/mixin.rb:219:in `print_version_notice'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.2.2-java/lib/logstash/config/mixin.rb:206:in `validate'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.2.2-java/lib/logstash/config/mixin.rb:102:in `config_init'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.2.2-java/lib/logstash/outputs/base.rb:52:in `initialize'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.2.2-java/lib/logstash/pipeline.rb:291:in `plugin'
(eval):38:in `initialize'
org/jruby/RubyKernel.java:1111:in `eval'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.2.2-java/lib/logstash/pipeline.rb:30:in `initialize'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.2.2-java/lib/logstash/agent.rb:109:in `execute'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.2.2-java/lib/logstash/runner.rb:91:in `run'
org/jruby/RubyProc.java:271:in `call'
/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.2.2-java/lib/logstash/runner.rb:96:in `run'
org/jruby/RubyProc.java:271:in `call'
/opt/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.21/lib/stud/task.rb:12:in `initialize'
```

and for now, have no idea where to search.

**Comment by talevy** [comment]
@sbeaupre, @Da-Wei's update changes made it into master. you can go ahead and try that now!

**Comment by sbeaupre** [comment]
@talevy you mean the update and upsert commit? That I saw and is also in the 1.0.7 release. What I was looking for was the scripting solution.

In the mean time, the error above was solved in 1.5.3, but I still need to try the changes of @Da-Wei with 1.0.7

**Comment by jsvd** [comment]
@talevy was this all implemented in #235 ? can we close this?

**Comment by talevy** [comment]
@jsvd, yes. looks like it
